File: android/src/test/java/com/google/android/libraries/privacy/ppn/internal/UptimeTrackerTest.java
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the );
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an  BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android.libraries.privacy.ppn.internal;

import static com.google.common.truth.Truth.assertThat;

import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.ZoneId;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

/** Unit tests for {@link PpnSettings}. */
@RunWith(RobolectricTestRunner.class)
public class UptimeTrackerTest {

  @Test
  public void tracker_defaultsToZero() {
    UptimeTracker tracker = new UptimeTracker();

    Duration duration = tracker.collectDuration(getStartClock());

    assertThat(duration.toMillis()).isEqualTo(0);
  }

  @Test
  public void startAndStop_hasDuration() {
    UptimeTracker tracker = new UptimeTracker();
    Clock startClock = getStartClock();
    Clock stopClock = Clock.offset(startClock, Duration.ofMillis(500));
    Clock measureClock = Clock.offset(stopClock, Duration.ofMillis(500));

    tracker.start(startClock);
    tracker.stop(stopClock);
    Duration duration = tracker.collectDuration(measureClock);

    assertThat(duration.toMillis()).isEqualTo(500);
  }

  @Test
  public void start_hasDuration() {
    UptimeTracker tracker = new UptimeTracker();
    Clock startClock = getStartClock();
    Clock measureClock = Clock.offset(startClock, Duration.ofMillis(500));

    tracker.start(startClock);
    Duration duration = tracker.collectDuration(measureClock);
    assertThat(duration.toMillis()).isEqualTo(500);
  }

  @Test
  public void collect_resetsDuration() {
    UptimeTracker tracker = new UptimeTracker();
    Clock startClock = getStartClock();
    Clock measureClock1 = Clock.offset(startClock, Duration.ofMillis(500));
    Clock measureClock2 = Clock.offset(measureClock1, Duration.ofMillis(750));

    tracker.start(startClock);
    Duration duration = tracker.collectDuration(measureClock1);
    assertThat(duration.toMillis()).isEqualTo(500);

    duration = tracker.collectDuration(measureClock2);
    assertThat(duration.toMillis()).isEqualTo(750);
  }

  @Test
  public void startTwice_isNoOp() {
    UptimeTracker tracker = new UptimeTracker();
    Clock startClock1 = getStartClock();
    Clock startClock2 = Clock.offset(startClock1, Duration.ofMillis(250));
    Clock measureClock = Clock.offset(startClock1, Duration.ofMillis(500));

    tracker.start(startClock1);
    tracker.start(startClock2);
    Duration duration = tracker.collectDuration(measureClock);
    assertThat(duration.toMillis()).isEqualTo(500);
  }

  @Test
  public void stopTwice_isNoOp() {
    UptimeTracker tracker = new UptimeTracker();
    Clock startClock = getStartClock();
    Clock stopClock1 = Clock.offset(startClock, Duration.ofMillis(500));
    Clock stopClock2 = Clock.offset(startClock, Duration.ofMillis(750));
    Clock measureClock = Clock.offset(stopClock2, Duration.ofMillis(2000));

    tracker.start(startClock);
    tracker.stop(stopClock1);
    tracker.stop(stopClock2);
    Duration duration = tracker.collectDuration(measureClock);

    assertThat(duration.toMillis()).isEqualTo(500);
  }

  @Test
  public void uptime_isCumulative() {
    UptimeTracker tracker = new UptimeTracker();
    Clock initClock = getStartClock();

    Clock startClock1 = Clock.offset(initClock, Duration.ofMillis(100));
    Clock stopClock1 = Clock.offset(startClock1, Duration.ofMillis(1));

    Clock startClock2 = Clock.offset(stopClock1, Duration.ofMillis(200));
    Clock stopClock2 = Clock.offset(startClock2, Duration.ofMillis(2));

    Clock startClock3 = Clock.offset(stopClock1, Duration.ofMillis(300));
    Clock measureClock = Clock.offset(startClock3, Duration.ofMillis(3));

    tracker.start(startClock1);
    tracker.stop(stopClock1);

    tracker.start(startClock2);
    tracker.stop(stopClock2);

    tracker.start(startClock3);
    Duration duration = tracker.collectDuration(measureClock);

    assertThat(duration.toMillis()).isEqualTo(6);
  }

  /** Returns a default fixed clock to use as the starting point for tests. */
  private Clock getStartClock() {
    Instant now = Instant.now();
    return Clock.fixed(now, ZoneId.systemDefault());
  }
}

File: krypton/desktop/windows/uptime_tracker_test.cc
/*
 * Copyright (C) 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "privacy/net/krypton/desktop/windows/uptime_tracker.h"

#include <memory>

#include "privacy/net/krypton/krypton_clock.h"
#include "testing/base/public/gunit.h"
#include "third_party/absl/time/time.h"

namespace privacy {
namespace krypton {
namespace windows {
namespace {

class UptimeTrackerTest : public ::testing::Test {
  void SetUp() override {
    clock_.SetNow(absl::UniversalEpoch() + absl::Minutes(5));
    uptime_tracker_ = std::make_unique<UptimeTracker>(&clock_);
  }

 protected:
  FakeClock clock_ = FakeClock(absl::UniversalEpoch() + absl::Minutes(5));
  std::unique_ptr<UptimeTracker> uptime_tracker_;
};

TEST_F(UptimeTrackerTest, Start_HasDuration) {
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(5));

  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::Minutes(5));
}

TEST_F(UptimeTrackerTest, StartAndStop_HasDuration) {
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(2));
  uptime_tracker_->Stop();
  clock_.AdvanceBy(absl::Minutes(3));

  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::Minutes(2));
}

TEST_F(UptimeTrackerTest, StopWithoutStart) {
  uptime_tracker_->Stop();
  clock_.AdvanceBy(absl::Minutes(5));
  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::ZeroDuration());
}

TEST_F(UptimeTrackerTest, CollectDuration_ResetDuration) {
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(10));
  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::Minutes(10));
  clock_.AdvanceBy(absl::Minutes(100));

  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::Minutes(100));
}

TEST_F(UptimeTrackerTest, StartTwice) {
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(5));
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(10));
  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::Minutes(15));
}

TEST_F(UptimeTrackerTest, StopTwice) {
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(5));
  uptime_tracker_->Stop();
  clock_.AdvanceBy(absl::Minutes(10));
  uptime_tracker_->Stop();
  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::Minutes(5));
}

TEST_F(UptimeTrackerTest, CollectDurationAccumulative) {
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(5));
  uptime_tracker_->Stop();
  clock_.AdvanceBy(absl::Minutes(10));
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(15));
  uptime_tracker_->Stop();
  clock_.AdvanceBy(absl::Minutes(20));
  uptime_tracker_->Start();
  clock_.AdvanceBy(absl::Minutes(25));
  uptime_tracker_->Stop();
  clock_.AdvanceBy(absl::Minutes(30));
  ASSERT_EQ(uptime_tracker_->CollectDuration(), absl::Minutes(45));
}

}  // namespace
}  // namespace windows
}  // namespace krypton
}  // namespace privacy
